Add bmm path for nested matmul

mikaylagawarecki opened the pull request:
Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, *_1, head_dim]`, `mat2 [N, n_heads, head_dim, *_2]`,

1. view them as 3D nested tensors `self [N * n_heads, *_1, head_dim]`, `mat2 [N * n_heads, head_dim, *_2]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, *_1, *_2]`,

This function is called in nested matmul if inputs are 4D, on CUDA and do not require gradients, a follow-up would be to add backward support

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #85857

mikaylagawarecki committed:
Add bmm fallback path for nested matmul

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm fallback path for nested matmul"

Registered a function that does this 

For two nested 4D inputs of shape `self [N, *, n_heads, head_dim]`, `mat2 [N, *, head_dim, n_heads]`,

1. view them as 3D nested tensors`self [sum(*), n_heads, head_dim]`, `mat2 [sum(*), head_dim, n_heads]`,
2. call bmm
3. view the output as `mat2 [N, *, n_heads, n_heads]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm fallback path for nested matmul"

### A backward kernel must be registered for this path if we need autograd support for it

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, head_dim, *]`, `mat2 [N, n_heads, *, head_dim]`,

1. view them as 3D nested tensors `self [N * n_heads, head_dim, *]`, `mat2 [N * n_heads, *, head_dim]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, head_dim, head_dim]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, head_dim, *]`, `mat2 [N, n_heads, *, head_dim]`,

1. view them as 3D nested tensors `self [N * n_heads, head_dim, *]`, `mat2 [N * n_heads, *, head_dim]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, head_dim, head_dim]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, head_dim, *]`, `mat2 [N, n_heads, *, head_dim]`,

1. view them as 3D nested tensors `self [N * n_heads, head_dim, *]`, `mat2 [N * n_heads, *, head_dim]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, head_dim, head_dim]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, *, head_dim]`, `mat2 [N, n_heads, head_dim, *]`,

1. view them as 3D nested tensors `self [N * n_heads, *, head_dim]`, `mat2 [N * n_heads, head_dim, *]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, *, *]`,




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -905,6 +1010,22 @@ Tensor matmul_nested(const Tensor& self, const Tensor& mat2) {
     mat2_dim_size);
+
+  // use bmm inference-only fast path for [N, n_heads, *, head_dim] [N, n_heads, head_dim, *]
+  if (self_dim == 4 && mat2_dim == 4) {
```
You can check here whether the inputs require gradients and then branch on that

cpuhrsch replied:
In particular you can check

```
torch.is_grad_enabled() and any(x.requires_grad for x in tensor_args):
```

From C++ that's [`GradMode::is_enabled()`](https://github.com/pytorch/pytorch/blob/a67621a6ca19b0c1423a7b136bfd90d8f04182fb/torch/csrc/autograd/init.cpp#L516) for torch.is_grad_enabled and `x.requires_grad()`

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, *_1, head_dim]`, `mat2 [N, n_heads, head_dim, *_2]`,

1. view them as 3D nested tensors `self [N * n_heads, *_1, head_dim]`, `mat2 [N * n_heads, head_dim, *_2]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, *_1, *_2]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, *_1, head_dim]`, `mat2 [N, n_heads, head_dim, *_2]`,

1. view them as 3D nested tensors `self [N * n_heads, *_1, head_dim]`, `mat2 [N * n_heads, head_dim, *_2]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, *_1, *_2]`,




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[WIP] Add bmm path for nested matmul"

### 1) A backward kernel must be registered for this path if we need autograd support for it and 2) we need to call this function in `matmul_nested` if xyz conditions hold

Registered a function that does this 

For two nested 4D inputs of shape `self [N, n_heads, *_1, head_dim]`, `mat2 [N, n_heads, head_dim, *_2]`,

1. view them as 3D nested tensors `self [N * n_heads, *_1, head_dim]`, `mat2 [N * n_heads, head_dim, *_2]`,
2. call bmm
3. view the output as `mat2 [N * n_heads, *_1, *_2]`,




[ghstack-poisoned]

mikaylagawarecki commented:
On AWS A100, for inputs of shape `[N, n_heads, *_1, head_dim], mat2 [N, n_heads, head_dim, *_2]` and patching https://github.com/pytorch/pytorch/pull/85894
after this change: `torch.float16 matmul() avg_time: 2.0176 ± 0.1706`
before this change: `torch.float16 matmul() avg_time: 25.6940 ± 7.0188`

<details>

<summary>benchmark script</summary>

```
from unittest import load_tests
import torch
import numpy as np
import time
from itertools import product

torch.manual_seed(5)
if torch.cuda.is_available():
    print("Using cuda: ", torch.cuda.is_available())
    device = 'cuda'
else:
    device = 'cpu'

def benchmark_torch_function(iters, f, *args, **kwargs):
    f(*args, **kwargs)
    if torch.cuda.is_available():
        torch.cuda.synchronize()
        start_event = torch.cuda.Event(enable_timing=True)
        end_event = torch.cuda.Event(enable_timing=True)
        start_event.record()
    else:
        t0 = time.time()
    for _ in range(iters):
        f(*args, **kwargs)
    if torch.cuda.is_available():
        end_event.record()
        torch.cuda.synchronize()
        return start_event.elapsed_time(end_event)
    else:
        return (time.time() - t0)

def fn1(nt1, nt2):
    y = nt1.matmul(nt2)

def zipf_sentence_lengths(alpha: float, batch_size: int, *, np_seed=0) -> np.ndarray:
    # sync generated lengths across runs
    np.random.seed(np_seed)
    sentence_lengths = np.empty(batch_size, dtype=int)
    for ibatch in range(batch_size):
        sentence_lengths[ibatch] = 1
        word = np.random.zipf(alpha)
        while word != 3 and word != 386 and word != 858:
            sentence_lengths[ibatch] += 1
            word = np.random.zipf(alpha)
    print(sentence_lengths)
    return sentence_lengths

alpha = 1.2
N = 512
E_q, E_k, E_v, E_total, E_out = 512, 512, 512, 512, 512
nheads = 8
E_head = E_total // nheads
iters = 100
dtype = torch.float16
# sentence_lengths = zipf_sentence_lengths(alpha, N)

# In tutorial we use matmul here attn_output = attn_weights.bmm(value)
# (N, nheads, L_t, L_s) x (N, nheads, L_s, E_head) -> (N, nheads, L_t, E_head)

# for padded version bmm is used attn_output = attn_weights.bmm(value)
# (N * nheads, L_t, L_s) x (N * nheads, L_s, E_head) -> (N * nheads, L_t, E_head)
# attn_output = attn_weights.bmm(value)

# L_t and L_s are jagged and are the sentence_lengths above

# generate 100 random zipf law data and run each for 100 iters
# return avg(avg_iters_i)
bmm_times = []
for i in range(100):
    print("i: ", i)
    sentence_lengths = zipf_sentence_lengths(alpha, N, np_seed=i)
    L_s = L_t = sentence_lengths
    max_len = sentence_lengths.max()
    fake_attn_weights = []
    fake_values = []
    for i in range(N):
        fake_attn_weights.append(torch.randn(nheads, L_t[i], L_s[i], dtype=dtype, device=device))
        fake_values.append(torch.randn(nheads, L_s[i], E_v, dtype=dtype, device=device))


    nested_fake_attn_weights = torch.nested.nested_tensor(fake_attn_weights)
    print(nested_fake_attn_weights.device, nested_fake_attn_weights._nested_tensor_size())
    nested_fake_value = torch.nested.nested_tensor(fake_values)
    print(nested_fake_value.device, nested_fake_value._nested_tensor_size())

    bmm_times.append(benchmark_torch_function(iters, fn1, nested_fake_attn_weights, nested_fake_value) / iters)
bmm_times = np.array(bmm_times)
avg_time = np.mean(bmm_times)
sd = np.std(bmm_times)
print(f'{dtype} matmul() avg_time: {avg_time:3.4f} \u00B1 {sd:3.4f}')
print('')
```
</details>

drisspg commented:
Doing some house keeping, was there a reason why we didn't merge this in? The speedups look good. If we want to merge in I think the only thing would maybe be adding `_matmul_with_bmm_nested` to nested namespace

mikaylagawarecki commented:
@drisspg this was landed by @cpuhrsch as part of #86856, closing as complete
